Gem adoption

sonalkr132 opened the pull request:
Related: #725
Adoption aims to facilitate gem ownership transfer between users, whether requested by gem owner or any rubygems.org user. ownership transfer is already possibly by `gem owner -a/r` command, however to quote bf4:
> Being able to make an 'ownership request' on rubygems.org and have current owners able to approve by clicking on a link or otherwise not having to go to the command line would reduce a lot of friction

RFC for said feature, talks of being able to flag gems by owners as 'looking for help' (`adoption`), following which other users can make 'adoption requests' (`adoption_request.opened?`). These requests can be either 'approved' (`adoption_request.approved?`) or 'rejected' (`adoption_request.closed?`). Differentiation between just looking for help and looking for owner/transfer can be communicated through `note` field of adoption. Approval of adoption request will only add requester as owner, and not remove any existing owner.
This PR allows adoption requests to be sent whether gem owner is actually looking for help (created adoption) or not. On our help site, number of tickets asking gem transfer for allegedly not-so-used gem is "too damn high". All we do on those tickets is add owner of gem to the ticket and hope they reply. Allowing adoption request being sent from site will automate some of our work. Also, I hope it will further discussion about ways of dealing with AWOL owners. Argument that an active user could have better use of a gem namespace currently occupied by an empty gem, with no releases in 5 years, under 2000 total downloads and whose owner has been unresponsive for over a year seems justifiable to me. Perhaps we can open adoption requests for upvote by other users and given gem's popularity, an upvote threshold could approve adoption. My numbers are entirely random. Coming to agreement on what constitutes an AWOL user and a not-so-used gem is no doubt going to be strenuous. I would like to think, this PR is a step in that direction.

## Todo (this PR)
- [x] integration tests
- [x] mobile view
- [x] internationalization
- [x] recent adoptions and adoption request approvals page
- [x] badge for gems looking for adoption

## Todo (future PRs)
* pagination in adoption*s* view
* rate limit on adoption applications and gem transfers per user.
* search adoptions
* gem page for adoption applications approvals history

## More todo
These are very broad and I am not going to work on them, more discussion/agreement is needed. Probably it will also mean we finalize our terms of use (#1336). I will open an issue with my suggestions when I reach here. I need to build a more-informed opinion.
* approval of adoption requests on basis of gem's last version release, total and monthly downloads, owner's unresponsiveness, presence of useful code in package (a lot of gems are just empty :rotating_light: ).
* allow users to approve adoption request by voting
* ability to retire gems

[:computer: shots](https://gist.github.com/sonalkr132/66262091bd812258e84ea4ae5fbe21e9)

---
I had started working on this before npm/event-stream situation. Above was my pitch which still holds, except it doesn't talk of what does ownership transfer means to users of the gem. How can we improve trust of users in code they are installing? This is not the first time this question has come up, and I would say there is fair consensus on [TUF](https://theupdateframework.github.io/):
> We plan to implement TUF on top of RubyGems in the future, and eventually deprecate the (not really secure) current gem signing system. In the meantime, treat signed gems with a large grain of salt. ([source](https://github.com/rubygems/rubygems.org/issues/1226#issuecomment-200949477))

Problem with TUF as put by tarcieri is: "TUF is, well, tough" :weight_lifting_man: . Last time I took it up, I ran out of (my) time before I could contribute anything useful. It is still on my TODO but would it be considered blocker for this PR?
Gem adoption in a way makes ownership transfer more explicit than what is already possible with `gem owner -a`. Gem adoption discussion mentions of safely retiring a gem and same comes up in https://github.com/rubygems/rubygems/issues/2496.

See also:  [rubygems/adoption-center](https://github.com/rubygems/adoption-center)

olleolleolle commented on app/views/adoptions/_default.html.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="t-body">
+  <p>Owner(s) of <%= @rubygem.name %> are not looking for maintainers. You can still request an adoption if
+  you feel gem namespace could be of greater service with your ownership and you can convince
```
```suggestion
  you feel this gem could be of greater service with your ownership and you can convince
```

olleolleolle commented on app/views/adoptions/_default.html.erb:
```diff
@@ -0,0 +1,7 @@
+  <p>Owner(s) of <%= @rubygem.name %> are not looking for maintainers. You can still request an adoption if
+  you feel gem namespace could be of greater service with your ownership and you can convince
+  existing owner(s) of the gem of the same.
+  We will notify owner(s) of <%= @rubygem.name %> with your adoption note, you can fill below.
```
```suggestion
  We will notify owner(s) of <%= @rubygem.name %> with your adoption note, which you can fill out below.
```

olleolleolle commented on app/views/adoptions/_request.html.erb:
```diff
@@ -0,0 +1,13 @@
+<% if adoption %>
+  <h3 class="t-list__heading"> You have requested adoption for <%= @rubygem.name %>:</h3>
```
```suggestion
  <h3 class="t-list__heading"> You have requested adoption of <%= @rubygem.name %>:</h3>
```

olleolleolle commented on app/views/adoptions/_request.html.erb:
```diff
@@ -0,0 +1,13 @@
+  <%= render partial: "adoptions/show", locals: { adoption: adoption } %>
+<% elsif signed_in? %>
+  <div class="gem_intro">
+    <h3 class="t-list__heading">Request adoption :</h3>
```
```suggestion
    <h3 class="t-list__heading">Request adoption:</h3>
```

sonalkr132 replied:
are you sure there will be a space before colon?
https://english.stackexchange.com/questions/67394/spaces-around-a-colon

bf4 commented:
@sonalkr132 @olleolleolle fwiw, here's notes I wrote up on 2016 from a discussion with @qrush that's been sitting in the Bundler slack 

Kudos for taking this on!

> Here's my notes with a bit of interp; I think we discussed mostly 
prioritizing the ability for something to ask to be an owner and 
letting an owner approve just by clicking a link in an email

```
RubyGem has_one OwnershipRequest
OwnersRequest
- description   [Request Message]
- contact_email [EmailAddress]
- closed [bool] # or destroy?
On Gem page, a logged in gem owner
sees in 'admin' section (above the 'links') section
click 'request owners'
enters a request description and a contact email
and clicks create
On Gem page, a logged in gem owner
see in 'admin' sections 
click 'edit owners request'
can modify any details or click 'close request'
and clicks update
On Gem page, for get with open Owners Request
any user sees a banner at the top with the body of the
ownership request and the contact email
RubyGem belongs_to OwnershipApplication
- user_id [ Integer]
- body # request message [String] max 255 char
- owner_id [Integer]
- ruby_gem_id [Integer]
- timestamps
On Gem page, a logged in User
Sees in the 'links' section
click request ownership
an OwnershipApplication is created for that user_id and with the message
owners gets email with 'authorize' link
owner clicks link
link is at /ownerships/approve/:sgid
link uses MessageVerifier sgid to verify request
ownership application is updated with the approving owner_id
requesting user is added as an owner to request gem
On Gem page with banner, a logged in user....?
```


>  the tl;dr as I see it is that for some reason, getting someone to `gem owner benjamins -a davids@davidsgems.tld` can stall or turn into a help ticket someone needs to deal with.  This proposal turns that into `please click here`

sonalkr132 commented:
@bf4 
My PR checks most boxes. Let me explain the differences.
> OwnershipRequest, OwnershipApplication

IMHO, `Adoption` better communicates **transfer** of gem/namespace than `OwnershipRequest`. I am not as happy with `AdoptionRequest`, perhaps `AdoptionApplication`? @olleolleolle what do you think?
`AdoptionRequest` has enum state as owners will be able to closing/rejecting requests and deleting someone else's record on that action seems wrong. Note that, this PR will allows adoption request to be sent regardless of fact that owner has

sonalkr132 commented:
Thank you for your input. @olleolleolle  I have updated your suggestions in i18n commit. I hope you found email bodies presentable as you didn't leave any suggestion on them.
@bf4 I have updated my PR as per your spec, except for a few cases:
> OwnershipRequest and OwnershipApplication

IMHO, Adoption and AdoptionApplication better communicate *transfer* of gem/namespace.
> contact_email [EmailAddress]
> click 'edit ownership request'

Adoption belongs to user. this is not needed. I would consider edit a *nice to have* too.

> any user sees a banner at the top with the body of the..
> body # request message [String] max 255 char

I will add a small github button like tag next to gem name where adoption exists. I think we should not limit characters of body/note, as both adoption and adoption requests are better when detailed with responsibilities and acceptance terms (check screenshots?).
> owners gets email with 'authorize' link .. /ownerships/approve/:sgid

If we are to store approver (owner_id/approved_id), we would need to create a token per owner per adoption application(one more relation). Previously, we had issue of email token being leaked as referrer url and they didn't even have any expiry time back then. I get the idea of making things easier but I think it's best if we avoid email tokens. This PR, sends note of adoption application along with link to gem adoptions path (where owner can approve/close applications).
We can set up approval by replying to email thread if @dwradcliffe  thinks its something we can/should do.

Lastly, AdoptionApplication also has enum status field, as owners can close/reject adoption application and we shouldn't delete application record for that action.

sonalkr132 commented:
I have added a sitewide page for adoptions index and a tag on gem (seeking adoption) show page.
This seems good to merge if there are no objections. There are a few left over things like pagination and email styling, which I intend to cover in follow up PRs.

<img src="https://user-images.githubusercontent.com/7680662/52711236-513d4000-2fb7-11e9-8c99-730dde31a7af.gif" width="400">

ioquatix commented:
I really like this idea. Is it deployed yet?

I'm wondering if this forms part of a wider discussion about how to handle gems which are abandoned by their author.

Namespace is a limited resource and there are a ton of gems which were published in 2010-2012 which only ever had a handful of `0.x.y` releases, < 10,000 downloads and are tying up valuable namespace which is making it hard for new contributors. Is there a good solution to this problem?

ioquatix commented:
Just dumping some ideas for further discussion:

## How to decide if gems are "unmaintained"

### Is the gem still in use?

Can we track monthly downloads? If so, we have a pretty good statistic on whether a gem is in use or not.

- Are reverse dependencies still active?

### Is the gem stable or in development?

Gems which have a "stable" release should have a higher threshold for adoption.

- If the gem has a 1.x release, we can consider it to be stable.
- If the gem only has 0.x releases, we can consider it to be in development.

### Is the source code still available?

- Validation on whether the source code is still available, e.g. follow the homepage link, does it give a 404 error? (tons of old/unmaintained gems = 404)

This could also be a warning sent to users - e.g. request that the URL is updated.

### Is the user still active?

If users are no longer active, and the gem is no longer being maintained, this is a strong signal that someone else could step in and maintain the gem or reuse the namespace.

- Validation on whether the user is responding to email: email all Rubygem authors once a year, perhaps with a summary of all their gems, and also ask them to confirm their account - maybe also a good opportunity to get people to update to 2FA.

## Logic

Here is some pseudo code. Feel free to play around with the ideas.

```ruby
class Gem
	def in_use?
		if monthly_downloads.last > 1000
			return true
		else
			reverse_dependencies.any?(&:in_use?)
		end
	end
	
	def maintained?
		author.logged_in_recently? && updated_at > 3.years.ago
	end
	
	def valid_metadata?
		response = get(homepage_url)
		
		!response.not_found?
	end
	
	def stable?
		version >= "1.0.0"
	end
	
	def adoption_threshold
		if stable? && maintained?
			return 6.years
		else
			return 3.years
		end
	end
	
	def automatic_adoption?
		last_released_at < adoption_threshold.ago
	end
end
```

We could also add a lock flag for gems which are critical, or part of stdlib, etc.

ioquatix commented:
I had one more idea about this, which would be an interesting thing I could do in isolation: upon downloading the gem and installing it (and dependencies), does the code load and do tests run?

There are a couple of ways this could work: using the declared minimum version of Ruby, or testing on all current rubies. It would give a good baseline as to whether gems were actually functioning. Given a gem that's unmaintained, not stable, and non-functional, is there an obligation by RubyGems to continue hosting it?

What could be interesting, is to add an API for tagging gems. If this was the case, I could, for example, build a local test server and actually do this, and then tag gems with the appropriate metadata.

ioquatix commented:
Where are we at with this? Is it good to merge? Do we need to review it?

sonalkr132 commented:
Hi @ioquatix ,
I am sorry about the delay in response here. It has been delayed because of a couple of reasons.

When we enabled gem push notification, the issue of spamming of gem push notification by adding random users was pointed out. Adding a web UI option, along with email notification for gem owner -a/r also felt more urgent than this.
bf4's comment suggesting the use of `ownership application` and `ownership request` also grew on me over time. It has a lot less room for interpretation than `adoption`.

The idea of allowing `ownership application` **without** the gem having an existing `ownership request` was not part of the original adoption RFC. I included it only to automate the process of contacting owners of inactive gems for transfers (done through help.rubygems.org until now). Thought of users replacing gems with malicious content after adoption makes me deeply concerned. Breached accounts make things bad as it is; we can't afford to make it worse. One of the ideas we suggest in RFC is adding an option for the existing owner of the gem to yank all current versions and lock the namespace for a given period before `ownership application` approval. Broken build is way better scenario than a compromised host.

I am hoping to get this done in GSoC 2020. Working with one of the applicants, we have opened [an RFC](https://github.com/rubygems/rfcs/pull/25) for the same. It addresses most of the concerns I mentioned above. For example, we will limit `ownership application` to gems by download count and updated_at time. I am hoping with old inactive gems, we will have a smaller risk of things going south and the impact contained.

It would great if you could go over it and leave clarifications/suggestions.

----
> I'm wondering if this forms part of a wider discussion about how to handle gems which are abandoned by their author.

Yes, I am hoping to automate gem transfers all the way. Some of your ideas are great, like checking if the gem is in use (total and weekly downloads, last version release) and checking if the gem has **any** code at all (a good number are just empty templates). I would also add the reputation of the user requesting adoption on our site as one of the easily quantifiable criteria. The user being inactive (no response to `ownership application`) would be a precondition for all of it.
Others are a bit controversial/impractical. We never had a requirement about version numbers (semvar) or sidebar links. If you think these can be useful, users will need a strong justification for making the switch now.

>  It would give a good baseline as to whether gems were actually functioning.
> is there an obligation by RubyGems to continue hosting it?

We understand the frustration you feel when you come up with an excellent name for your gem and find it already taken Deciding the functionality of a gem can be a controversial topic, and I wouldn't recommend it. It would also come at the cost of increased infra spending.
We want to maximize the freedom of **all** users and be as less interventionist as possible.

Thank you for your patience. Please feel free to open an RFC for dealing with AWOL owners if you are interested.

ioquatix commented:
I think all this work and idea is great. Thanks for the update.

You asked for feedback. Well I agree with your perspective.

However, I have one other thought.

Maybe it's time `rubygems.org` has some kind of namespace.

Per user/org namespace would be very great for security and coordinating releases.

I don't know if I need to explain it or not. Most of the function is already there (using gem source do ... end).

Let me know if you'd like to have a chat some time, I am in the bundler slack.

simi commented:
@ioquatix I was thinking about the namespacing (similar to npm) as well. I think I have posted my ideas to some issue already. :eyes:

ioquatix commented:
I don't really have the mental space to provide a huge amount of feedback right now but I trust everyone here to understand the gist of it and if anyone has time to implement it, that would be awesome.

sonalkr132 commented:
> Per user/org namespace would be very great for security and coordinating releases.

From what I understand we didn't agree that this is a good idea even if frequently requested, reasons being:
- unnecessary fragmentation in community
- too much work (for something already supported with `git` source)

These two links[[1](https://quaran.to/blog/2009/10/09/on-gem-forking/)][[2](https://github.com/rubygems/rubygems.org/issues/1226#issuecomment-200463298)] are a good place to start if you would like judge this on your own. Please feel free to submit an RFC if you think it should be reconsidered.

We didn't have any issue with few gem transfers we have done through help site so far, still we need to carefully think about security implications of adoption. Limit on downloads for `ownership application` is there, so that If things do go south, only a small number of users will be affected. I think in beta release we will start with 10_000 downloads and re-evaluate need to quarantine the gem namespace before transfer.

ioquatix commented:
Fair enough, I trust people have thought about it and I trust the conclusion is considered.

simi commented:
@sonalkr132 those links are pretty old discussions. A lot things have changed meanwhile. My main concerns are about global namespace naming and also organization based ownership (access to org. means access to all gems in there).

Anyway since it is not really related to gem adoption at all, I can think about that more and try to write down RFC with my ideas.

sonalkr132 commented:
>  pretty old discussions.

Just about the time, I was added as member here, I was told that rg.org doesn't really need any shiny new feature/idea, stability is the only thing that is important. I don't disagree.
Picking old projects where there is already enough consensus seemed like a justifiable exception to me and there are enough such projects on rg.org to keep me occupied so far. I haven't spent much time evaluating viability of anything new.

This doesn't mean we absolutely will never entertain anything new. namespace has been an all time popular request, an RFC collecting all the idea would be awesome. Thanks.

I am closing this now, I hope I have answered your queries satisfactorily.
